### PR TITLE
NativeAOT: Inline tls access on linux/x64

### PR DESCRIPTION
**Pull request by kunalspathak**
Before:
```asm
  G_M41298_IG02:  ;; offset=0x0001
         E800000000           call     CORINFO_HELP_READYTORUN_THREADSTATIC_BASE
         488D0D00000000       lea      rcx, gword ptr [(reloc 0x40000000004210a8)]      ; '"a"'
  
         48894858             mov      gword ptr [rax+0x58], rcx
                                                ;; size=16 bbWeight=1 PerfScore 2.50
```

After:
```asm
  G_M41298_IG02:  ;; offset=0x0008
         66                   data16   
         488D3D00000000       lea      rdi, [(reloc 0x4000000000421230)]
         66                   data16   
         66                   data16   
         48E800000000         call     <unknown method>
         488B18               mov      rbx, gword ptr [rax]
         4885DB               test     rbx, rbx
         7412                 je       SHORT G_M41298_IG05
                                                ;; size=24 bbWeight=1 PerfScore 5.50
  G_M41298_IG03:  ;; offset=0x0020
         488D0500000000       lea      rax, gword ptr [(reloc 0x4000000000421278)]      ; '"a"'
  
         48894358             mov      gword ptr [rbx+0x58], rax
                                                ;; size=11 bbWeight=1 PerfScore 1.50
  G_M41298_IG04:  ;; offset=0x002B
         4883C408             add      rsp, 8
         5B                   pop      rbx
         5D                   pop      rbp
         C3                   ret      
                                                ;; size=7 bbWeight=1 PerfScore 2.25
  G_M41298_IG05:  ;; offset=0x0032
         488BF8               mov      rdi, rax
         488D0500000000       lea      rax, [(reloc 0x4000000000421238)]
         FFD0                 call     rax
         488BD8               mov      rbx, rax
         EBDD                 jmp      SHORT G_M41298_IG03
                                                ;; size=17 bbWeight=0 PerfScore 0.00
  
  ; Total bytes of code 67, prolog size 8, PerfScore 12.75, instruction count 23, allocated bytes for code 67 (MethodHash=14305ead) for method Program:Test1() (FullOpts)
```

The following sequence is later [patched by the linker](https://github.com/llvm/llvm-project/blob/90c326b198080c5c208f62f6755d54d7b69b291d/lld/ELF/Arch/X86_64.cpp#L459-L472) as seen below:

```asm
         66                   data16   
         488D3D00000000       lea      rdi, [(reloc 0x4000000000421230)]
         66                   data16   
         66                   data16   
         48E800000000         call     <unknown method>

```

The end result in this code:

image

### CODE REVIEW

**Comment by kunalspathak** [comment]
@jkotas @VSadov @MichalStrehovsky @BruceForstall PTAL
@dotnet/jit-contrib

**Comment by kunalspathak** [comment]
@Mihubot

**Comment by kunalspathak** [comment]
@MihaZupan - is this way to trigger the jit-diff?

**Comment by MihaZupan** [comment]
Oh I see, the check I had was case-sensitive for the bot name. Fixed now
@MihuBot

**Comment by VSadov** [comment]
Codegen looks nice. No more calls when acquiring a lock!

```gas
System.Collections.Concurrent.Tests`System.Threading.Lock__EnterAndGetCurrentThreadId:
    0x5555564cae70 <+0>:   pushq  %rbp
    0x5555564cae71 <+1>:   pushq  %r15
    0x5555564cae73 <+3>:   pushq  %rbx
    0x5555564cae74 <+4>:   subq   $0x10, %rsp
    0x5555564cae78 <+8>:   leaq   0x20(%rsp), %rbp
    0x5555564cae7d <+13>:  movq   %rdi, %rbx

// start fetching managed TLS
->  0x5555564cae80 <+16>:  movq   %fs:0x0, %rax
    0x5555564cae89 <+25>:  leaq   -0x108(%rax), %rax
    0x5555564cae90 <+32>:  movq   %rax, %rdi
    0x5555564cae93 <+35>:  movq   (%rdi), %r15              ; here r15 points to tlsRoot object
    0x5555564cae96 <+38>:  testq  %r15, %r15                ; check if the root is not null (otherwise call helper)
    0x5555564cae99 <+41>:  je     0x5555564caee0            ; <+112> at Lock.cs:66
    0x5555564cae9b <+43>:  movl   0x88(%r15), %edx

// here edx has the thread ID - in 7 instructions! 

// the rest of the Lock code follows
// check if the states of the lock word and the thread ID permit fast path (typically it is the case)
    0x5555564caea2 <+50>:  testl  %edx, %edx
    0x5555564caea4 <+52>:  je     0x5555564caec9            ; <+89> at Lock.cs:66
    0x5555564caea6 <+54>:  movl   0x14(%rbx), %eax
    0x5555564caea9 <+57>:  movl   %eax, -0x14(%rbp)
    0x5555564caeac <+60>:  testb  $0x3, %al
    0x5555564caeae <+62>:  jne    0x5555564caec9            ; <+89> at Lock.cs:66
    0x5555564caeb0 <+64>:  leal   0x1(%rax), %edi
    0x5555564caeb3 <+67>:  leaq   0x14(%rbx), %rsi

// acquire the lock
    0x5555564caeb7 <+71>:  lock
    0x5555564caeb8 <+72>:  cmpxchgl %edi, (%rsi)

    0x5555564caebb <+75>:  movl   -0x14(%rbp), %edi
    0x5555564caebe <+78>:  cmpl   %edi, %eax
    0x5555564caec0 <+80>:  jne    0x5555564caec9            ; <+89> at Lock.cs:66
    0x5555564caec2 <+82>:  movl   %edx, 0x10(%rbx)
    0x5555564caec5 <+85>:  movl   %edx, %eax
    0x5555564caec7 <+87>:  jmp    0x5555564caed6            ; <+102> at Lock.cs:69
    0x5555564caec9 <+89>:  movq   %rbx, %rdi
    0x5555564caecc <+92>:  movl   $0xffffffff, %esi         ; imm = 0xFFFFFFFF
    0x5555564caed1 <+97>:  callq  0x5555564cb0a0            ; System.Threading.Lock__TryEnterSlow_0 at Lock.cs:297
    0x5555564caed6 <+102>: nop
    0x5555564caed7 <+103>: addq   $0x10, %rsp
    0x5555564caedb <+107>: popq   %rbx
    0x5555564caedc <+108>: popq   %r15
    0x5555564caede <+110>: popq   %rbp
    0x5555564caedf <+111>: retq
    0x5555564caee0 <+112>: leaq   0x79959(%rip), %rax       ; Internal.Runtime.ThreadStatics__GetInlinedThreadStaticBaseSlow at ThreadStatics.cs:40
    0x5555564caee7 <+119>: callq  *%rax
    0x5555564caee9 <+121>: movq   %rax, %r15
    0x5555564caeec <+124>: jmp    0x5555564cae9b            ; <+43> at Lock.cs:66
```

**Comment by VSadov** [comment]
I've run a few tests in a loop in various configurations (Checked/Release, ServerGC,...) and did not see any crashes or anomalies. 

Can't comment on the code change details, but the end result looks as expected!

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!!

File: src/coreclr/inc/corinfo.h [hunk 1]
```diff
@@ -1749,6 +1749,7 @@ struct CORINFO_THREAD_STATIC_INFO_NATIVEAOT
     CORINFO_CONST_LOOKUP tlsRootObject;
     CORINFO_CONST_LOOKUP tlsIndexObject;
     CORINFO_CONST_LOOKUP threadStaticBaseSlow;
+    CORINFO_CONST_LOOKUP tlsGetAddrFtnPtr;
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:1752]: This changing JIT/EE interface. Do you need to update JIT-EE interface GUID and PMI?

File: src/coreclr/jit/helperexpansion.cpp [hunk 2]
```diff
@@ -609,100 +612,128 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
...
+        //      data16 data16
+        //      call     _tls_get_addr
+        //
+        // This sequence along with `data16` prefix is expected by the linker so it
+        // will patch these with TLS access.
+        GenTree* tls_get_addr_val =
+            gtNewIconHandleNode((size_t)threadStaticInfo.tlsGetAddrFtnPtr.handle, GTF_ICON_FTN_ADDR);
+        tls_get_addr_val->SetContained();
+
+        // GenTreeCall* tlsRefCall = gtNewCallNode(CT_ tls_get_addr_val, TYP_I_IMPL);
```
Comment (BruceForstall) [inline src/coreclr/jit/helperexpansion.cpp:636]: unneeded?

File: src/coreclr/jit/helperexpansion.cpp [hunk 3]
```diff
@@ -609,100 +612,128 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
...
+        GenTreeCall* tlsRefCall = gtNewIndCallNode(tls_get_addr_val, TYP_I_IMPL);
+        tlsRefCall->gtFlags |= GTF_TLS_GET_ADDR;
+        // //
```
Comment (BruceForstall) [inline src/coreclr/jit/helperexpansion.cpp:639]: extraneous comments?

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/helperexpansion.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/helperexpansion.cpp:636]: I will do this in a follow-up PR

**Pull request merged**